source/server/listener_impl.h

```diff
@@ -18,6 +19,69 @@ namespace Server {
 
 class ListenerManagerImpl;
 
+class ListenSocketFactoryImplBase : public Network::ListenSocketFactory,
+                                    protected Logger::Loggable<Logger::Id::config> {
+public:
+  ListenSocketFactoryImplBase(ListenerComponentFactory& factory,
+                              Network::Address::InstanceConstSharedPtr local_address,
+                              Network::Address::SocketType socket_type,
+                              const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
+                              const std::string& listener_name);
+
+  // Network::ListenSocketFactory
+  Network::Address::SocketType socketType() const override { return socket_type_; }
+  const Network::Address::InstanceConstSharedPtr& localAddress() const override {
+    return local_address_;
+  }
+
+protected:
+  Network::SocketSharedPtr createListenSocketAndApplyOptions();
+  void setLocalAddress(Network::Address::InstanceConstSharedPtr local_address);
+
+private:
+  ListenerComponentFactory& factory_;
+  // Initially, its port number might be 0. Once a socket is created, its port
+  // will be set to the binding port.
+  Network::Address::InstanceConstSharedPtr local_address_;
+  Network::Address::SocketType socket_type_;
+  const Network::Socket::OptionsSharedPtr options_;
+  bool bind_to_port_;
+  const std::string& listener_name_;
+};
+
+class TcpListenSocketFactory : public ListenSocketFactoryImplBase {
+public:
+  TcpListenSocketFactory(ListenerComponentFactory& factory,
+                         Network::Address::InstanceConstSharedPtr local_address,
+                         const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
+                         const std::string& listener_name);
+
+  // Network::ListenSocketFactory
+  // If |socket_| is nullptr, create a new socket for it. Otherwise, always return |socket_|.
+  Network::SocketSharedPtr createListenSocket() override;
+  absl::optional<std::reference_wrapper<Network::Socket>> sharedSocket() const override {
+    return *socket_;
+  }
+
+private:
+  // This will be shared across all worker threads if not setting SO_REUSEPORT.
```

Comment:
Can you rephrase a bit just to make it clear this is not supported today? (reformat to TODO, etc.)